sessions/2018/october/Session_15.md

```diff
@@ -0,0 +1,324 @@
+}
+```
+
+That response is showing a book with `bookID: 0` was created.
+
+After it, in order to access by HTTP GET to see our newly created book (or any other) we can go to: <http://localhost:8080/api/book/0> replacing the last digit for the desired `bookID`. Our code is prepared to answer to any inexistent `bookID` returning a null JSON object, for example:
+
+The request <http://localhost:8080/api/book/1000>
+
+Returns:
+```json
+{
+bookID: null,
+title: null,
+year: null
+}
+```
+
+#### API desing concepts
+
+Is possible to serve Frontend Web content in Spring:
+<https://spring.io/guides/gs/serving-web-content/>
+<https://www.baeldung.com/spring-request-response-body>
+
+An example to work with HTML inside Spring is using JSP. This is very hard to work with, is very strict and not good for error tracking: https://docs.spring.io/spring/docs/3.2.x/spring-framework-reference/html/view.html
+
+While this is possible to serve frontend content, the aproach of separate interests must be followed, since this allow more flexibility and portability. So instead to use a full stack application in a box, we can delegate an API to do the logic tasks while we can use whatever view engine or framework we want in order to access the API.
+
+This is not the same as microservices, is just to separate the view from the API logic.
+
+This is neither the same as Progressive Web Apps (PWA), since in these kind of applications, the content is being loaded and served from the browser self, so even without an internet connection, the application works and can serve the content has been already loaded and sync when a internet connection is available again. By the other side, in the API approach, a network connection is mandatory for the correct operation of the API and its components.
+
+In the API approach, one of the things to have present is that APIs are decoupled from the view. So we can use and switch them between different platforms. Some usual components to achieve this approach are:
+
+- API using some backend programming language such as Java for example.
+- SPA (Single Page Application) is the frontend in which we can use JS, CSS, HTML5 or some JS framework such as React, Angular or Vue since the're some of the most popular SPA options right now.
+
+
+#### Dependency injection concepts
+
+A simplification and better practice to use dependency injection in Spring was talked about. We were changing from field injection towards constructor injection.
+
+##### Field injection
+
+Field injection allows to link resources, so for example in the next code snippet, a controller is asking to Spring to `@Autowire` the `BookService` class from its pool of registered services:
+
+```java
+import org.springframework.beans.factory.annotation.Autowired;
+
+@RestController
+@RequestMapping("/api/book")
+public class BookController {
+
+    @Autowired
+    private BookService bookService;
+    // ...
+}
+```
+
+__NOTE__: Field injection is the best choise for JUnit tests.
```

Comment:
In particular, it's the only place where it's acceptable. This is because constructor injection cannot be used (JUnit requires the test class to have a default constructor)